typings/objection/index.d.ts

```diff
@@ -911,101 +860,101 @@ declare namespace Objection {
     | Buffer
     | Raw
     | Literal;
-  type ColumnRef = string | Raw | Reference | QueryBuilder<any>;
-  type TableName = string | Raw | Reference | QueryBuilder<any>;
-
-  interface QueryInterface<T> {
-    select: Select<T>;
-    as: As<T>;
-    columns: Select<T>;
-    column: Select<T>;
-    from: Table<T>;
-    into: Table<T>;
-    table: Table<T>;
-    distinct: Distinct<T>;
+  type ColumnRef = string | Raw | Reference | QueryBuilder<any, any[]>;
+  type TableName = string | Raw | Reference | QueryBuilder<any, any[]>;
+
+  interface QueryInterface<QM, RM, RV> {
```

Comment:
I know `QueryInterface` was not added in this commit, but I have always wondered what't the purpose of it? Only one class ever implements it. Couldn't we just move these methods over to `QueryBuilder` and eliminate the whole interface?

Reply:
Are the QueryInterface methods all borrowed from Knex, while the QueryBuilder methods added by Objection? If so, it might help with maintenance to keep them separate. But then we might want to rename QueryInterface to make that clearer.

I do find the names QueryBuilder and QueryBuilderBase confusing. Both names suggest that they're the top-most query builder interface, but QueryBuilder extends QueryBuilderBase. At least in this new merging of query builders, every QueryBuilder instance implements QueryBuilderBase, so we could merge these two interfaces.

Reply:
If you can pull it off and not break existing tests, go for it: less is more.

Reply:
I got answers. The QueryBuilder and QueryBuilderBase types may correspond to classes of those names in the implementation. If these types do indeed correspond to these classes, we may want to keep the names the same to facility maintenance.

OTH, as a person perusing just the public API -- including the typings -- I think I'd prefer calling QueryBuilderBase and QueryBuilder by the names of AbstractQueryBuilder and QueryBuilder, to tell me their roles and relationship at a glance.

Also, I was unable to merge these two interfaces because the return types of `throwIfNotFound()` cannot be generalized more strictly than `any`.

I did however move `returning()` and `runAfter()` to QueryBuilderBase. I think `runAfter()` may have originally been query-builder-specific in order to map the output of the query builder to the input of `runAfter()`. But because `runAfter()` callbacks can be chained, we can't hardcode its input type. I changed it to `any`.